### PR TITLE
enable official plugins via command-line option

### LINKED ISSUE #10: Do not hardcode plugins
Instead of hardcoding the plugins here (see below) allow them to be toggled in command-line arguments.

https://github.com/srid/tailwind-haskell/blob/d3b1e27aa384bf295d05d1d716425d32b90e3e79/src/Web/Tailwind.hs#L116-L121

### PR DESCRIPTION
**Pull request by rubenmoor**
Resolves #10

### CODE REVIEW

**Comment by rubenmoor** [comment]
https://github.com/srid/tailwind-haskell/issues/10

File: src/Web/Tailwind.hs [hunk 1]
```diff
@@ -43,23 +44,34 @@ import UnliftIO (MonadUnliftIO, finally)
...
+    _tailwindConfigPlugins :: [OfficialPlugin]
   }
   deriving (Generic)
-  deriving
-    (ToJSON)
-    via CustomJSON
-          '[ FieldLabelModifier
-               '[StripPrefix "_tailwindConfig", CamelToSnake]
-           ]
-          TailwindConfig
```
Comment (srid) [inline src/Web/Tailwind.hs:62]: Why was this removed?

File: src/Web/Tailwind.hs [hunk 2]
```diff
@@ -43,23 +44,34 @@ import UnliftIO (MonadUnliftIO, finally)
...
+
+instance Text.Show.Show OfficialPlugin where
+  show = \case
+    PluginTypography  -> "typography"
+    PluginForms       -> "forms"
+    PluginLineClamp   -> "line-clamp"
+    PluginAspectRatio -> "aspect-ratio"
+
+mkStrPlugin :: OfficialPlugin -> Text
+mkStrPlugin plugin = "require('@tailwindcss/" <> Text.pack (show plugin) <> "')"
```
Comment (srid) [inline src/Web/Tailwind.hs:63]: Let's inline this where it gets used.

File: src/Web/Tailwind.hs [hunk 3]
```diff
@@ -43,23 +44,34 @@ import UnliftIO (MonadUnliftIO, finally)
 import UnliftIO.Directory (removeFile)
 import UnliftIO.Process (callProcess)
 import UnliftIO.Temporary (withSystemTempFile)
+import qualified Data.Text as Text
+
+data OfficialPlugin
```
Comment (srid) [inline src/Web/Tailwind.hs:49]: Let's call it `data Plugin`

File: exe/Main.hs [hunk 4]
```diff
@@ -24,6 +53,11 @@ cliParser = do
   content <- NE.some (argument str (metavar "SOURCES..."))
   output <- strOption (long "output" <> short 'o' <> metavar "OUTPUT" <> value "tailwind.css")
   mode <- modeParser
+  bPluginTypography  <- switch (long "plugin-typography"   <> help "enable official typography plugin")
+  bPluginForms       <- switch (long "plugin-forms"        <> help "enable official forms plugin")
+  bPluginLineClamp   <- switch (long "plugin-line-clamp"   <> help "enable official line-clamp plugin")
+  bPluginAspectRatio <- switch (long "plugin-aspect-ratio" <> help "enable official aspect ratio plugin")
+  bAllPlugins        <- switch (long "all-plugins" <> short 'a' <> help "enable all official plugins")
```
Comment (srid) [inline exe/Main.hs:60]: The new CLI disables all plugins by default, no? Existing behaviour (all plugins enabled) shouldn't change.

Also, I feel like we want to have one option that takes comma-separated values: `--plugins="typography,forms,line-clamp,aspect-ratio"`.

**Review by srid — CHANGES_REQUESTED** [review]
Thanks! Reviewed:

File: src/Web/Tailwind.hs (hunk 1 above)
Comment (rubenmoor) [reply src/Web/Tailwind.hs:62]: We can do something like this:

```haskell
data Plugin
  = Typography
  | Forms
  | LineClamp
  | AspectRatio
  deriving (Generic)
  deriving (ToJSON, FromJSON)
    via CustomJSON '[ ConstructorTagModifier '[ CamelToKebab ] ] Plugin

data TailwindConfig = TailwindConfig
  { -- | List of source patterns that reference CSS classes
    _tailwindConfigContent :: [FilePath],
    _tailwindConfigPlugins :: [Plugin],
    _tailwindConfigTheme   :: Aeson.Object
  }
  deriving (Generic)
  deriving
    (ToJSON)
    via PrefixedSnake "_tailwindConfig" TailwindConfig

instance Text.Show.Show TailwindConfig where
  show (decodeUtf8 . encode -> config) =
    toString
      [text|
      module.exports = ${config}
      |]
```

But `tailwind.config.js` isn't really JSON, it contains JavaScript.
So the `ToJSON` instance isn't actually that helpful at all.

I don't have a good idea, though. The `[text| ...]` template works fine, though.
We can patch the file together.

----

What is your opinion on having the `theme` object as `Aeson.Object` in `TailwindConfig`?
I added a command-line argument to add some file "theme.json", which is then JSON-decoded.

File: exe/Main.hs (hunk 4 above)
Comment (rubenmoor) [reply exe/Main.hs:60]: > The new CLI disables all plugins by default, no? Existing behaviour (all plugins enabled) shouldn't change.

got it

> Also, I feel like we want to have one option that takes comma-separated values: --plugins="typography,forms,line-clamp,aspect-ratio".

Yes, I can certainly do that. There isn't a standard way of passing list-like stuff via command-line, unfortunately.
The interface that you suggest is very clear. It requires custom reader instances for `data Plugin` and will bloat the code a bit more.

**Comment by rubenmoor** [comment]
Now, theme customizations are included via `TailwindConfigTheme` and for the command-line tool via file.
The file isn't treated as Json however. Arbitrary Javascript is allowed.
Therefore, the file is appended to the (temporary) tailwind.config.js **as is**.

File: src/Web/Tailwind.hs (hunk 1 above)
Comment (srid) [reply src/Web/Tailwind.hs:62]: Oh, okay. I thought it is used in Emanote, but if not - we can get rid of the json instance.

> What is your opinion on having the theme object as Aeson.Object in TailwindConfig?
I added a command-line argument to add some file "theme.json", which is then JSON-decoded.

Useful, but let's do it in separate PR, while keeping this PR focused.

File: exe/Main.hs [hunk 5]
```diff
@@ -48,3 +107,9 @@ main = do
         ( fullDesc
             <> progDesc "Run Tailwind"
         )
+    readPlugins txt =
+      let ePlugins = for (Text.splitOn "," txt) \strPlugin ->
+            case readMaybe $ Text.unpack $ Text.toLower $ Text.strip strPlugin of
+              Just plugin -> pure plugin
+              Nothing     -> Left $ "Could not parse " <> strPlugin
+      in  either error id ePlugins
```
Comment (srid) [inline exe/Main.hs:115]: You can actually do the parsing (and error handling) as part of CLI argument parsing. Example: https://github.com/srid/rib/blob/db26bcde09364849683736bc1dc30acdff5d6cc0/rib-core/src/Rib/Cli.hs#L108

**Comment by srid** [comment]
> Now, theme customizations are included via `TailwindConfigTheme` and for the command-line tool via file. The file isn't treated as Json however. Arbitrary Javascript is allowed. Therefore, the file is appended to the (temporary) tailwind.config.js **as is**.

Please have this PR add `--plugins` only. It is easier to review that way. You can add other features as separate PR. I would also appreciate as minimal a diff as possible (refactorings can always be done as separate PRs).

**Comment by srid** [comment]
(I pushed a merge commit in this PR just so I can test my CI setup; cf. https://github.com/srid/nixos-config/pull/34)

**Comment by rubenmoor** [comment]
closing in favor of https://github.com/srid/tailwind-haskell/pull/13

File: src/Web/Tailwind.hs (hunk 1 above)
Comment (srid) [reply src/Web/Tailwind.hs:62]: > we can get rid of the json instance.

I was wrong, sorry. See https://github.com/srid/tailwind-haskell/pull/13#discussion_r1142842486
